platform/search/src/GenericSearchProvider.js

```diff
@@ -0,0 +1,267 @@
+                var ids = [],
+                    id;
+                
+                // If we have the results from a search 
+                if (event.data.request === 'search') {
+                    // Convert the ids given from the web worker into domain objects
+                    for (id in event.data.results) {
+                        ids.push(id);
+                    }
+                    objectService.getObjects(ids).then(function (objects) {
+                        var searchResults = [],
+                            id;
+                        
+                        // Create searchResult objects
+                        for (id in objects) {
+                            searchResults.push({
+                                object: objects[id],
+                                id: id,
+                                score: event.data.results[id]
+                            });
+                        }
+                        
+                        // Resove the promise corresponding to this 
+                        pendingQueries[event.data.timestamp].resolve({
+                            hits: searchResults,
+                            total: event.data.total,
+                            timedOut: event.data.timedOut
+                        });
+                    });
+                }
+            }
+            
+            worker.onmessage = handleResponse;
+            
+            // Helper function for getItems(). Indexes the tree.
+            function indexItems(nodes) {
+                nodes.forEach(function (node) {
+                    var id = node && node.getId && node.getId();
+                    
+                    // If we have not yet indexed this item, index it 
+                    if (!indexed[id]) {
+                        // Index each item with the web worker
+                        indexItem(node);
+                        indexed[id] = true;
+                        
+                        if (node.hasCapability && node.hasCapability('composition')) {
+                            // This node has children
+                            
+                            // Make sure that this is async, so doesn't block up page
+                            $timeout(function () {
+                                // Get the children... 
+                                node.useCapability('composition').then(function (children) {
+                                    $timeout(function () {
+                                        // ... then index the children
+                                        if (children.constructor === Array) {
+                                            indexItems(children);
+                                        } else {
+                                            indexItems([children]);
+                                        }
+                                    }, 100);
+                                });
+                            }, 100);
```

Comment:
1. How many packets/objects end up in the index after this work is completed?  That will help estimate how much lag is okay.
2. Instead of a 100ms timeout, try a 0ms timeout.  It works roughly the same-- any timeout allows the UI to update.  
3. The function executed in each timeout will block the UI as long as it's executing, so each one of those functions should complete as quick as possible too.  Can you isolate which lines of code are taking the most CPU time?
4. In the case of objects that support mutation: is it possible that you are attaching multiple listeners to the same object?  Might want to ensure that's not happening.

Pete